Simplify API hierarchy of HTTP and gRPC client builders

chemicL opened the pull request:
Motivation:

The current hierarchy of HTTP and gRPC client builders is complicated
and lacks alignment between both stacks. The HTTP hierarchy in
particular carries along some historical context which was recently
updated to make the `Partitioned`- and `MultiAddress`- `HttpClientBuilder`s
independent of the API useful for `SingleAddressHttpClientBuilder` as
keeping them consistent was problematic and led to the introduction of
`SingleAddressInitializer` mechanism in both cases. After recent
untangling of particular clients' dependency on those shared APIs it's
now time to simplify the hierarchy, which this change attempts.

Modifications:

- Removed `BaseGrpcClientBuilder`,
- Removed `SingleAddressGrpcClientBuilder`,
- Removed `BaseSingleAddressHttpClientBuilder`,
- Moved methods of `HttpClientBuilder` into
  `SingleAddressHttpClientBuilder` to highlight their meaning in the
  context of a client that is dedicated to a particular address. This
  change left the `HttpClientBuilder` interface empty,
- Merged `ExecutionContextAwareHttpBuilder` and
  `HttpClientBuildFinalizer` into `HttpClientBuilder` as those
  package-private interfaces were useful in the untangling process. Now
  they're not required any more and can be incorporated into
  `HttpClientBuilder` which solely represents the responsibilities that
  those interfaces encapsulated,
- Reverted `MultiAddressHttpClientBuilder` and
  `PartitionedHttpClientBuilder` to abstract classes instead of
  interfaces to remain consistent with the server builders and the gRPC
  counterparts.

Result:

The API hierarchy is simplified, uses similar concepts for different
levels across the HTTP and gRPC stack and the more complex multi- and
partitioned- client builders have a common base interface that doesn't
carry unnecessary mechanisms from the single address use case. At the
same time, for most usages where the `HttpClients` and `GrpcClients`
factories are used, there would be no action required from the users.

bondolo reviewed (commented):
Great simplification. I don't feel confident to approve though

idelpivnitskiy commented on servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java:
```diff
@@ -50,72 +58,235 @@
+     * <p>
+     * The order of execution of these filters are in order of append. If 3 filters are added as follows:
+     * <pre>
+     *     builder.append(filter1).append(filter2).append(filter3)
```
To be consistent with HTTP, please use the real method name here: `appendConnectionFactoryFilter`

idelpivnitskiy commented on servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java:
```diff
@@ -50,72 +58,235 @@
+     * <p>
+     * The order of execution of these filters are in order of append. If 3 filters are added as follows:
+     * <pre>
+     *     builder.append(filter1).append(filter2).append(filter3)
```
To be consistent with HTTP, please use the real method name here: `appendConnectionFilter`

idelpivnitskiy commented on servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java:
```diff
@@ -50,72 +58,235 @@
+     * <p>
+     * The order of execution of these filters are in order of append. If 3 filters are added as follows:
+     * <pre>
+     *     builder.append(filter1).append(filter2).append(filter3)
```
To be consistent with HTTP, please use the real method name here: `appendConnectionFilter`.

Same for other `append*` methods we have below (pre-existing, I can not comment there).

idelpivnitskiy commented on servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java:
```diff
@@ -50,72 +58,235 @@
+     * @param sslConfig The configuration to use.
+     * @return {@code this}.
+     */
     public abstract GrpcClientBuilder<U, R> sslConfig(ClientSslConfig sslConfig);
```
Consider adding `@see ClientSslConfigBuilder` tag

idelpivnitskiy commented on servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java:
```diff
@@ -50,72 +58,235 @@
+     * @return {@code this}
+     */
     public abstract GrpcClientBuilder<U, R> autoRetryStrategy(
             AutoRetryStrategyProvider autoRetryStrategyProvider);
```
Consider adding `@see DefaultAutoRetryStrategyProvider` tag

idelpivnitskiy commented on servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java:
```diff
@@ -50,72 +58,235 @@
+     * @param strategy {@link GrpcExecutionStrategy} to use.
+     * @return {@code this}.
+     */
     public abstract GrpcClientBuilder<U, R> executionStrategy(GrpcExecutionStrategy strategy);
```
Consider adding `@see GrpcExecutionStrategies` tag

idelpivnitskiy commented on servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java:
```diff
@@ -50,72 +58,235 @@
+     * @param protocols {@link HttpProtocolConfig} for each protocol that should be supported.
+     * @return {@code this}.
+     */
     public abstract GrpcClientBuilder<U, R> protocols(HttpProtocolConfig... protocols);
```
Consider adding `@see HttpProtocolConfigs` tag

chemicL replied:
For that to be visible, the grpc-api module would need to depend on http-netty. Do we want that?

idelpivnitskiy replied:
No, that was overlook, I thought it's in api

idelpivnitskiy commented on servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java:
```diff
@@ -50,72 +58,235 @@
+     *
+     * @param loadBalancerFactory {@link HttpLoadBalancerFactory} to create {@link LoadBalancer} instances.
+     * @return {@code this}.
+     */
```
Can not comment on line 381:
Interesting that we do not have any test or example for `buildMulti()`. We should revisit if it's useful or not.

idelpivnitskiy commented on servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpClientBuilder.java:
```diff
@@ -45,256 +26,63 @@
-     */
-    public abstract HttpClientBuilder<U, R, SDE> appendConnectionFactoryFilter(
-            ConnectionFactoryFilter<R, FilterableStreamingHttpConnection> factory);
+    HttpClientBuilder<U, R, SDE> executionStrategy(HttpExecutionStrategy strategy);
```
Consider adding `@see HttpExecutionStrategies` tag

idelpivnitskiy commented on servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java:
```diff
@@ -39,88 +49,255 @@
+     * @param protocols {@link HttpProtocolConfig} for each protocol that should be supported.
+     * @return {@code this}.
+     */
     public abstract SingleAddressHttpClientBuilder<U, R> protocols(HttpProtocolConfig... protocols);
```
Consider adding `@see HttpProtocolConfigs` tag

chemicL replied:
That would require the api module to depend on http-netty.

idelpivnitskiy commented on servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java:
```diff
@@ -39,88 +49,255 @@
+     * @return {@code this}
+     */
+    public abstract SingleAddressHttpClientBuilder<U, R> autoRetryStrategy(
+            AutoRetryStrategyProvider autoRetryStrategyProvider);
```
Consider adding `@see DefaultAutoRetryStrategyProvider` tag

idelpivnitskiy commented on servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java:
```diff
@@ -39,88 +49,255 @@
+     * @return {@code this}.
+     */
+    public abstract SingleAddressHttpClientBuilder<U, R> loadBalancerFactory(
+            HttpLoadBalancerFactory<R> loadBalancerFactory);
```
Comment for the line 307:
Consider adding `@see ClientSslConfigBuilder` tag

idelpivnitskiy commented on servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultPartitionedHttpClientBuilder.java:
```diff
@@ -62,7 +62,7 @@
 import static java.util.function.Function.identity;
 
-class DefaultPartitionedHttpClientBuilder<U, R> implements PartitionedHttpClientBuilder<U, R> {
+class DefaultPartitionedHttpClientBuilder<U, R> extends PartitionedHttpClientBuilder<U, R> {
```
This class can be `final`

idelpivnitskiy reviewed (approved):
Great simplification!
Minor suggestions, otherwise LGTM

tkountis reviewed (approved):
Nice work!!

chemicL commented:
Thanks @bondolo @idelpivnitskiy @tkountis for the review. I addressed most comments, adding changes where there's no explanation why it wouldn't work.
